kamon-kafka-clients/src/main/java/kamon/kafka/client/instrumentation/advisor/Advisors.java

```diff
@@ -48,23 +49,26 @@
         public static void onEnter(@Advice.Argument(value = 0, readOnly = false) ProducerRecord record,
                                    @Advice.Argument(value = 1, readOnly = false) Callback callback,
                                    @Advice.Local("scope") Storage.Scope scope) {
-
             val currentContext = Kamon.currentContext();
             val topic = record.topic() == null ? "kafka" : record.topic();
             val partition = record.partition() == null ? "unknown-partition" : record.partition().toString();
             val value = record.key() == null ? "unknown-key" : record.key().toString();
 
-            val span = Kamon.buildSpan("kafka.produce")
-                    .asChildOf(currentContext.get(Span.ContextKey()))
-                    .withMetricTag("span.kind", "producer")
-                    .withTag("kafka.key", value)
-                    .withTag("kafka.partition", partition)
-                    .withTag("kafka.topic", topic)
+            val span = Kamon.producerSpanBuilder("kafka.produce", "kafkaPublisher")
+                    .asChildOf(currentContext.get(Span.Key()))
+                    .tag("span.kind", "producer")
+                    .tag("kafka.key", value)
+                    .tag("kafka.partition", partition)
+                    .tag("kafka.topic", topic)
                     .start();
 
-            val ctx = currentContext.withKey(Span.ContextKey(), span);
+            val ctx = currentContext.withEntry(Span.Key(), span);
+
+            val out = new ByteArrayOutputStream();
+            // ugly, ugly, ugly ... :(
+            Kamon.defaultBinaryPropagation().write(ctx, kamon.context.BinaryPropagation$ByteStreamWriter$.MODULE$.of(out));
```

Comment:
yes, is really ugly, maybe we can put this in a scala class and expose a method more friendly for java side :(

Reply:
Extract into dedicated class